core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,420 @@
+        /**
+         * A copy of the current tombstones in the builder.
+         */
+        public List<Tombstone> tombstones() {
+            return new ArrayList<>(tombstones);
+        }
+
+        /**
+         * Add a deleted index to the list of tombstones in the cluster state.
+         * Returns the tombstone that was added.
+         */
+        public Tombstone addTombstone(final Index index) {
+            Tombstone tombstone = new Tombstone(index, System.currentTimeMillis());
+            tombstones.addLast(tombstone);
+            return tombstone;
+        }
+
+        /**
+         * Purge tombstone entries.  Returns the number of entries that were purged.
+         *
+         * Tombstones are purged if the number of tombstones in the list
+         * is greater than the input parameter of maximum allowed tombstones.
+         * Tombstones are purged until the list is equal to the maximum allowed.
+         */
+        public int purge(final int maxTombstones) {
+            final int numToPurge = tombstones().size() - maxTombstones;
+            for (int i = 0; i < numToPurge; i++) {
+                // start from the beginning, it has the oldest entries
+                tombstones.removeFirst();
+            }
+            return numToPurge > 0 ? numToPurge : 0;
+        }
+
+        public IndexGraveyard build() {
+            return new IndexGraveyard(tombstones);
+        }
+    }
+
+    /**
+     * A class representing a diff of two IndexGraveyard objects.
+     */
+    final public static class IndexGraveyardDiff implements Diff<IndexGraveyard> {
+
+        private final List<Tombstone> added;
+        private final List<Tombstone> removed;
+
+        IndexGraveyardDiff(final StreamInput in) throws IOException {
+            added = Collections.unmodifiableList(in.readList(Tombstone.PROTOTYPE));
+            removed = Collections.unmodifiableList(in.readList(Tombstone.PROTOTYPE));
+        }
+
+        IndexGraveyardDiff(final IndexGraveyard previous, final IndexGraveyard current) {
+            final List<Tombstone> previousTombstones = previous.tombstones;
+            final List<Tombstone> currentTombstones = current.tombstones;
+            final List<Tombstone> added = new ArrayList<>();
+            final List<Tombstone> removed = new ArrayList<>();
+            if (previousTombstones.isEmpty()) {
+                // nothing will have been removed, and all entries in current are new
+                added.addAll(currentTombstones);
+            } else if (currentTombstones.isEmpty()) {
+                // nothing will have been added, and all entries in previous are removed
+                removed.addAll(previousTombstones);
+            } else {
+                // look through the back, starting from the end, for added tombstones
+                final Tombstone lastAddedTombstone = previousTombstones.get(previousTombstones.size() - 1);
+                for (int i = currentTombstones.size() - 1; i >= 0; i--) {
```

Comment:
here we could use sublists again - just scan to the place you need. No need to reverse then.